osu.Game/Skinning/RulesetSkinProvidingContainer.cs

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using JetBrains.Annotations;
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Game.Beatmaps;
+using osu.Game.Rulesets;
+
+namespace osu.Game.Skinning
+{
+    /// <summary>
+    /// A type of <see cref="SkinProvidingContainer"/> that provides access to the beatmap skin and user skin,
+    /// each transformed with the ruleset's own skin transformer individually.
+    /// </summary>
+    public class RulesetSkinProvidingContainer : SkinProvidingContainer
+    {
+        protected readonly Ruleset Ruleset;
+        protected readonly IBeatmap Beatmap;
+
+        protected override Container<Drawable> Content { get; }
+
+        public RulesetSkinProvidingContainer(Ruleset ruleset, IBeatmap beatmap, [CanBeNull] ISkin beatmapSkin)
+        {
+            Ruleset = ruleset;
+            Beatmap = beatmap;
+
+            InternalChild = new BeatmapSkinProvidingContainer(GetRulesetTransformedSkin(beatmapSkin))
+            {
+                Child = Content = new Container
+                {
+                    RelativeSizeAxes = Axes.Both,
+                }
+            };
+        }
+
+        [Resolved]
+        private SkinManager skinManager { get; set; }
+
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            UpdateSkins();
+        }
+
+        protected override void OnSourceChanged()
+        {
+            UpdateSkins();
+            base.OnSourceChanged();
+        }
+
+        protected virtual void UpdateSkins()
+        {
+            SkinSources.Clear();
+
+            SkinSources.Add(GetRulesetTransformedSkin(skinManager.CurrentSkin.Value));
+
+            // TODO: we also want to return a DefaultLegacySkin here if the current *beatmap* is providing any skinned elements.
+            if (skinManager.CurrentSkin.Value is LegacySkin && skinManager.CurrentSkin.Value != skinManager.DefaultLegacySkin)
+                SkinSources.Add(GetRulesetTransformedSkin(skinManager.DefaultLegacySkin));
+
+            if (skinManager.CurrentSkin.Value != skinManager.DefaultSkin)
+                SkinSources.Add(GetRulesetTransformedSkin(skinManager.DefaultSkin));
```

Comment:
Can you give an example of what cases require this? I would hope that the default skin should never need to be looked up via a transformer, as they are meant to be used for legacy cases only.

Reply:
I'm still not sure how this whole system of ruleset skin transformers is done, but I think we may still want that logic so that rulesets can still add their own logic regardless of the skin. (`GetDrawableComponent` implementations for example)

Reply:
Not sure I follow. Can you provide an example? The whole transformer classes are there for legacy purposes, and documented as such. If you can think of a purpose which is not legacy related we need to ensure it can work without them.

Reply:
> If you can think of a purpose which is not legacy related we need to ensure it can work without them.

You mean that non-legacy rulesets will not rely on `GetDrawableComponent` calls for its specific implementations (like osu!/taiko/catch/mania rulesets do) and directly make `SkinReloadableDrawable` subclasses that would directly fetch the resources or otherwise?

If so, then I think I can understand now. I was just worried about how rulesets would be interacting with lazer skins after removing that.

Reply:
The idea is that we would be doing things similar to how the HUD is done. Skins would specify the full hierarchy, including which components and which bindable properties to set on those components. `GetDrawableComponent` would likely never be used.

We would probably look to remove it eventually, updating the legacy skins to use similar logic.